### PR TITLE
[IMP] let postgres eliminate duplicates

### LINKED ISSUE #5806: Why making ids unique by hand instead of using distinct?
In https://github.com/odoo/odoo/blob/8.0/openerp/models.py#L4641, we select a list of ids and uniquify them unconditionally afterwards in https://github.com/odoo/odoo/blob/8.0/openerp/models.py#L4652

Could you please explain the rationale behind doing it this way instead of saying `select distinct` above? Postgres should be faster doing this and the current way, we always run into problems when `limit != None`

### PR DESCRIPTION
**Pull request by hbrunn**
fixes #5806

### CODE REVIEW

File: openerp/models.py [hunk 1]
```diff
@@ -4539,18 +4547,26 @@ def _search(self, cr, user, args, offset=0, limit=None, order=None, context=None
 
         limit_str = limit and ' limit %d' % limit or ''
         offset_str = offset and ' offset %d' % offset or ''
-        query_str = 'SELECT "%s".id FROM ' % self._table + from_clause + where_str + order_by + limit_str + offset_str
+        query_str = 'SELECT %s "%s".id FROM ' % (
+            'DISTINCT ON (' +
```
Comment (nbessi) [inline openerp/models.py]: That's tickle a little. It is hard to read. You should split that pice of code.

**Comment by nbessi** [comment]
@hbrunn did you make some benchmarking? I'm curious to see the effective gain of this patch

File: openerp/models.py [hunk 2]
```diff
@@ -4499,9 +4498,18 @@ def _generate_order_by(self, order_spec, query):
...
-                order_by_clause = ",".join(order_by_elements)
+            return order_by_elements
+        return []
 
+    def _generate_order_by(self, order_spec, query):
+        """
+        Attempt to consruct an appropriate ORDER BY clause based on order_spec, which must be
+        a comma-separated list of valid field names, optionally followed by an ASC or DESC direction.
+
+        :raise" except_orm in case order_spec is malformed
```
Comment (nbessi) [inline openerp/models.py]: I guess a `:` is missing after `raise`  I think correct usage is `:raises ExceptionType: docstring`

**Comment by hbrunn** [comment]
@nbessi I don't think this will gain a lot of performance. It should be a little faster, but that's not the point of this patch. The point is that searching with a limit on a field with _auto_join can yield less than `limit` results if there are duplicates, which confuses the pager (and is simply wrong)

**Comment by hbrunn** [comment]
I did sign the CLA :-( https://github.com/odoo/odoo/commit/91e5456fafc16b90359086a58e795dda4089031c

**Comment by nbessi** [comment]
@hbrunn we are on the same wavelength. I was just curious :wink:

**Comment by xmo-odoo** [comment]
@hbrunn 

> I did sign the CLA :-( 91e5456

I haven't checked the CLA script but I'm guessing it does its CLA check in the merge target branch, and 91e5456 hasn't been merged in master yet.

/cc @odony for confirmation/infirmation, and review

**Comment by hbrunn** [comment]
@xmo-odoo thank for picking that up!
@nbessi you made me curious too. I did some benchmarking with the unit tests, but there no difference comes out. Then I figured that the tests are not very search intensive, so I wrote a little script (https://gist.github.com/hbrunn/9362775fea147ed5709f) to benchmark searching only. Results look good so far (every line = one pass):

```
master: 6.40sec master-select_distinct: 4.48sec (70.00%) master-select_distinct_if_necessary: 4.49sec (70.00%)
master: 6.75sec master-select_distinct: 3.91sec (57.00%) master-select_distinct_if_necessary: 4.25sec (62.00%)
master: 6.72sec master-select_distinct: 4.20sec (62.00%) master-select_distinct_if_necessary: 3.91sec (58.00%)
```

(the third branch involved is https://github.com/hbrunn/OCB/tree/master-select_distinct_if_necessary, of which I'm quite surprised that is does not seem to improve anything)

**Comment by nbessi** [comment]
@hbrunn interesting results this is better than I expected thanks :smile:

**Comment by hbrunn** [comment]
take those results with a grain of salt, the performance boost grows with the amount of results returned. I assume most practical queries have less results and thus take less advantage from this change

**Comment by hbrunn** [comment]
The customer was unhappy that counting still is wrong in this case, so I added a distinct selection here too. I didn't do any measurements yet though, possibly we want https://github.com/hbrunn/OCB/tree/master-select_distinct_if_necessary after all for counting

**Comment by pedrobaeza** [comment]
@hbrunn is this still relevant?

**Pull request closed without merging**